### PR TITLE
Implementing `TextProperty` and `StringProperty` in `ndb`.

### PR DESCRIPTION
**Pull request by dhermes**

### CODE REVIEW

File: ndb/src/google/cloud/ndb/model.py [hunk 1]
```diff
@@ -1833,17 +1833,122 @@ def _db_get_value(self, v, unused_p):
...
+    A :class:`TextProperty` is intended for values of unlimited length, hence
+    is **not** indexed by default. A :class:`TextProperty` can be indexed via:
+
+    .. code-block:: python
+
+        class Item(ndb.Model):
+            description = ndb.TextProperty(indexed=True)
+            ...
+
+    but that is not the intended usage. If indexed text is desired, a
```
Comment (dhermes) [inline ndb/src/google/cloud/ndb/model.py]: @chrisrossi Do you think we should make this explicit by removing `indexed` from the constructor and making

```python
@property
def _indexed(self):
    return False
```

Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Done

File: ndb/src/google/cloud/ndb/model.py [hunk 2]
```diff
@@ -1833,10 +1833,94 @@ def _db_get_value(self, v, unused_p):
...
+        """
+        if isinstance(value, str):
+            return value.encode("utf-8")
+
+    def _from_base_type(self, value):
+        """Convert a value from the "base" value type for this property.
+
+        .. note::
+
+            Older versions of ``ndb`` could write non-UTF-8 ``TEXT``
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: :+1:

File: ndb/src/google/cloud/ndb/model.py [hunk 3]
```diff
@@ -1833,17 +1833,157 @@ def _db_get_value(self, v, unused_p):
...
+            NotImplementedError: Always. This method is virtual.
+        """
         raise NotImplementedError
 
 
 class StringProperty(TextProperty):
+    """An indexed property that contains UTF-8 encoded text values.
+
+    This is nearly identical to :class:`TextProperty`, but is indexed. Values
+    must be at most 1500 bytes (when encoded from UTF-8 to bytes).
```
Comment (chrisrossi) [inline ndb/src/google/cloud/ndb/model.py]: Encoding *to* UTF-8, not from.

File: ndb/src/google/cloud/ndb/model.py [hunk 4]
```diff
@@ -1833,10 +1833,94 @@ def _db_get_value(self, v, unused_p):
...
+                    "Expected valid UTF-8, got {!r}".format(value)
+                )
+        elif isinstance(value, str):
+            length = len(value.encode("utf-8"))
+        else:
+            raise exceptions.BadValueError(
+                "Expected string, got {!r}".format(value)
+            )
+
+        if self._indexed and length > _MAX_STRING_LENGTH:
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: I guess that's technically more accurate. I don't have a strong feeling about it.

File: ndb/src/google/cloud/ndb/model.py [hunk 5]
```diff
@@ -1833,10 +1833,94 @@ def _db_get_value(self, v, unused_p):
...
+        Returns:
+            Optional[str]: The converted value. If ``value`` is a a valid UTF-8
+            encoded :class:`bytes` string, this will return the decoded
+            :class:`str` corresponding to it. Otherwise, it will return
+            :data:`None`.
+        """
+        if isinstance(value, bytes):
+            try:
+                return value.decode("utf-8")
+            except UnicodeError:
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: Well, UnicodeError is a super-class of UnicodeDecodeError, right? Raising specific and catching general make sense to me.

[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Yes it is a super-class. :+1:

File: ndb/src/google/cloud/ndb/model.py (hunk 3 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Yeah I didn't like that sentence when I wrote it. I really meant from `str`, so I'll write that.

File: ndb/src/google/cloud/ndb/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: :+1:, will change.

**Pull request merged**